tests/st/calico_upgrade/test_calico_upgrade.py

```diff
@@ -0,0 +1,206 @@
+        rcu = calicoupgrade("dry-run")
+        logger.debug("INFO: calico-upgrade dry-run should return 0.")
+        rcu.assert_no_error()
+
+        ready_output = get_value_etcdv2("/calico/v1/Ready")
+        logger.debug("INFO: etcdv2 Ready flag is set to:\n%s" % ready_output)
+        assert ready_output == "true"
+
+        dry_run_output = open("/code/calico-upgrade-report/convertednames", 'rU')
+        dry_run_output_lines = dry_run_output.readlines()
+        dry_run_output.close()
+        logger.debug(
+            "INFO: calico-upgrade dry-run convertednames output:\n%s" % dry_run_output_lines)
+
+        rcu = calicoupgrade("start")
+        logger.debug("INFO: calico-upgrade start should return 0.")
+        rcu.assert_no_error()
+
+        ready_output = get_value_etcdv2("/calico/v1/Ready")
+        logger.debug("INFO: etcdv2 Ready flag is set to:\n%s" % ready_output)
+        assert ready_output == "false"
+
+        datastore_ready_rc = _get_ready_etcdv3()
+        assert datastore_ready_rc is False
+
+        start_output = open("/code/calico-upgrade-report/convertednames", 'rU')
+        start_output_lines = start_output.readlines()
+        start_output.close()
+        logger.debug("INFO: calico-upgrade start convertednames output:\n%s" % start_output_lines)
+
+        if dry_run_output_lines == start_output_lines:
+            logger.debug("INFO: calico-upgrade dry-run and calico-upgrade start output are equal")
+        else:
+            assert dry_run_output_lines == start_output_lines
+
+        rcc = calicoctl("convert", data=testdata)
+
+        rcc.assert_no_error()
+        parsed_output = yaml.safe_load(rcc.output)
+        converted_data = clean_calico_data(parsed_output)
+        logger.debug("INFO: converted data to v3\n%s" % converted_data)
+        original_resource = rcc
+
+        rcc = calicoctl("get %s %s -o yaml" % (converted_data['kind'], name(converted_data)))
+        logger.debug("INFO: calicoctl (v3) get - after calico-upgrade start: \n%s" % rcc.output)
+
+        # Comparison here needs to be against cleaned versions of data to remove Creation Timestamp
+        logger.debug("Comparing 'get'ted output with original converted yaml")
+        cleaned_output = yaml.safe_dump(
+            clean_calico_data(
+                yaml.safe_load(rcc.output),
+                extra_keys_to_remove=['projectcalico.org/orchestrator', 'namespace']
+            )
+        )
+        original_resource.assert_data(cleaned_output)
+
+        rcu = calicoupgrade("complete")
+        logger.debug("INFO: calico-upgrade complete should return 0.")
+        rcu.assert_no_error()
+
+        ready_output = get_value_etcdv2("/calico/v1/Ready")
+        logger.debug("INFO: etcdv2 Ready flag is set to:\n%s" % ready_output)
+        assert ready_output == "false"
+
+        datastore_ready_rc = _get_ready_etcdv3()
+        assert datastore_ready_rc is True
+    else:
+        rcu = calicoupgrade("dry-run")
```

Comment:
I think the dry run should be a completely separate test.
The upgrade test then doesn't need to first perform a dry run (since it's tested separately).

Both sets of tests can use the same input data.
The upgrade test should also be tested with bad data.  It should fail on the `start` command for the same reasons that the `dry-run` command fails

Reply:
Actually - maybe the separate test should be to run dry-run and start with the same set of data and check that they:
(a) both fail or both pass as expected
(b) the output from both is the same

Reply:
unless I'm talking rubbish and the `start` command only outputs stuff when it's successful.  I can't remember the behavior now.

Either way, we should at the very least test that `start` fails with bad input data.

Reply:
addressed with: 
A separate test running dry-run and start with the same set of data and check that they:
(a) both fail or both pass as expected
(b) the output from both is the same